rfc/rfc-0000-maven-artifacts-support.md

```diff
@@ -0,0 +1,160 @@
+# Meta
+[meta]: #meta
+- Name: Maven Artifacts Support
+- Start Date: 2022-03-30
+- Author(s): @emmjohnson
+- Status: Draft
+- RFC Pull Request: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Modify `owner.spec.source` to allow for specifying a JAR file.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+  - Today we support workloads being built from git and image source. Many developers in the field also need support for JARs.
+- What use cases does it support?
+  - Maven
+- What is the expected outcome?
+  - To enable more types of workloads.
+
+# What it is
+[what-it-is]: #what-it-is
+We need to modify the Owner specs to create an unambiguous way for specifying a JAR file. There are two proposed options, the goal is to get community buy-in on a solution before ratifying and the RFC will then be modified to suggest the agreed upon spec.
+
+This option allows for the developer to know what is necessary for a maven workload by examining the spec.
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: Workload
+metadata:
+  name: my-workload
+spec:
+  source:
+      maven: 
+        url: https://my-artifactory/my-projects/my-app.jar
```

Comment:
We searched around on an artifactory and I'm not sure I saw all of the fields that you listed. Would love it if you had an example you could point me to.

Reply:
> We searched around on an artifactory and I'm not sure I saw all of the fields that you listed. Would love it if you had an example you could point me to.

https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#dependency-management

Reply:
> Thanks @scothis. Definitely outside of my element here. We did modify the example to drop the `.jar`. So you are suggesting we need something like:
> 
> ```
> source:
>   maven:
>     groupid:
>     artifactid:
>     url:
>     index:
>     version:
>     config:
> ```

My suggestion is to not change the Workload resource until we have a sense of what the resource being stamped out looks like. I would expect something like this, but that's possibly insufficient based on the how the stamped resource defines the model. Most prominently this model assumes the set of repositories to resolve against are defined somewhere else.

```yaml
spec:
  source:
    maven:
      groupId:
      artifactId:
      version:
      type:
```

Reply:
At the very least that spec is missing the repository location. I was hoping there was a Resourceful URL that properly represented an artifact.

Reply:
Here is what concourse was using for its parameters for maven - https://github.com/nulldriver/maven-resource.

Reply:
It was my intent that the new `MavenRepository` resource which will be stamped out by `ClusterSourceTemplate` will have input fields very similar to what the Maven Resource for Concourse (linked by @emmjohnson in the above comment) uses:

* `url`: Required. The location of the repository.
* `snapshot_url`: Optional. The location of the snapshot repository.
* `artifact`: Required. The artifact coordinates in the form of groupId:artifactId:type[:classifier]
* `username`: Optional. Username for accessing an authenticated repository.
* `password`: Optional. Password for accessing an authenticated repository.
* `skip_cert_check`: Optional. If set to true, will ignore all certificate errors when accessing an SSL repository. NOTE: this will supersede the repository_cert configuration if also specified.
* `repository_cert`: Optional. CA/server certificate to use when accessing an SSL repository.

The `Workload` resource should take the same inputs (or maybe some platform-operator-subset of them could be pushed off to the supply chain configuration, as appropriate).

Reply:
Strong opinion: Those secrets need to be secrets and not params
Weak opinion: the artifact is probably easier to build as 3 fields as per POM layout.

Reply:
> perhaps?
> 
> ```yaml
> spec:
>   source:
>     maven:
>       repository: 
>       groupId: 
>       artifactId:
>       version:
>       type:
> ```
> 
> where repository is just the repository url?

While an artifact exists in a repository, it's uncommon to look in exactly one repository for that artifact. Even for the same groupId/artifactId, it's very common to split releases, milestones and snapshots into distinct repositories.

Reply:
@scothis What is type here? Maven? Or something else?

Reply:
> @scothis What is `spec.source.maven.repository.type` here?

It's the "type" of packaging for the artifact. By default the type is `jar`, but could also be `pom`, `maven-plugin`, `ejb`, `war`, `ear`, or `rar` (perhaps more, I'm not sure). They use different names for the same concept in different contexts...

https://maven.apache.org/pom.html#packaging

https://maven.apache.org/pom.html#dependencies

Reply:
reading https://maven.apache.org/pom.html#dependencies, `classifier` could also be important

Reply:
I have updated the design document again.  The new suggested specification for the `MavenResource` fields are now:

```
spec:
  artifact:
    id: groupId:artifactId:type[:classifier] # required
    versionFilter: ">=3.1.0-rc.1 <3.2.0"     # optional
  repository:
    url: https://my-maven-repository.com     # required
    authentication:
      secretRef: my-authn-secret             # optional
  transport:                                 # optional
    skipSslCheck: true or false              # optional
    caCertificate:
      secretRef: my-ca-certificate           # optional
  interval: 5m                               # optional, default 1m
```

This specification accommodates the `repository.type` and `classifier` fields in the `artifact.id` field.  The secrets have been broken out and will be specified with `secretRef` fields.  I figured we would need one of them for the application-level authentication (i.e.: user name and password for the repository) and another secret for transport-level authentication (e.g. CA certificate).

When it comes to dividing parameters between the `Workload` and the rest of the supply chain, maybe only the fields under `artifact` need to go under the `Workload` and the rest could be handled by the supply chain?

Reply:
@rszumlakowski Not sure the language that the MavenResource would be written in.  However, if it were a native java application it could make use of [DefaultArtifact](https://github.com/apache/maven-resolver/blob/master/maven-resolver-api/src/main/java/org/eclipse/aether/artifact/DefaultArtifact.java) which comes from the Aether library which is what is used by maven internally to resolve dependneices.  I contributed to a similar [capability in Spinnaker](https://github.com/spinnaker/clouddriver/blob/master/clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/maven/MavenArtifactCredentials.java#L51) many years ago.  Maven [version resultion](https://github.com/spinnaker/clouddriver/blob/master/clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/maven/MavenArtifactCredentials.java#L148) worked similar to the `spec.artifact.id` that you are referencing above.

Reply:
I agree with accounting for authentication and the possibility that the maven repository would have a custom ca.  The downside of using the artifact id notation `groupId:artifactId:type[:classifier]` is it does not carry with it the repository the artifact might resolve like in the case of a container image registry reference.  In the case of an ImageRepository reference, we get the k8s native capability for configuring PullSecrets to use by configuring registry credentials in the developer namespace.  Since this is a CRD I would propose that there be a custom secret of type MavenRepositorySecret.

```
apiVersion: v1
kind: Secret
metadata:
  name: MyMavenRepositorySecret
type: maven.org/MavenRepositorySecret
data:
  uri: https://my-maven-repository.com
  username: repo-username
  password: repo-password
  skipSslCheck: true or false
  caCertificate: |
    <certificate details go here>
```

Then the workload could reference the MavenRepository:

```
spec:
  artifact:
    id: groupId:artifactId:type[:classifier] # required
    versionFilter: ">=3.1.0-rc.1 <3.2.0"     # optional
    repository: 
      secretRef: MyMavenRepositorySecret
```

If a repository.secretRef is not provided then it could default to say Maven Central or a default secret of type maven.org/MavenRepositorySecret in the namespace.  If more than one is declared then something would have to error out.  Taking this approach would eliminate the need for the developer to need to configure all of the repository metadata for each workload.  Given a namespace could many artifacts declared that could come from the same maven repository the amount of boiler plate per workload could be cut down by using this proposal.

Reply:
@xyloman : Unfortunately, the controller that will be managing the `MavenRepository` resource will be written in Golang, like the rest of Cartographer and related TAP software, so we won't be able to use those handy Java APIs.

Also, I've update the `MavenRepository` specification to break up the `artifactId` field into individual fields, like this:

```yaml
spec:
  artifact:
    groupId: com.my.place                    # required
    artifactId: my-application               # required
    packaging: jar                           # required
    classifier: real-classy                  # optional
    semver: ">=3.1.0-rc.1 <3.2.0"            # optional
  repository:
    url: https://my-maven-repository.com     # required
    authentication:
      secretRef: my-authn-secret             # optional
  transport:                                 # optional
    skipSslCheck: true or false              # optional
    caCertificate:
      secretRef: my-ca-certificate           # optional
  interval: 5m                               # optional, default 1m
```

I like the idea of the custom `MavenRepositorySecret` secret type.  Are there any thoughts from the Cartographer team?  @jwntrs @emmjohnson ?  I'm not sure if we will have time to implement it for the first release, though.

Reply:
@rszumlakowski I agree with breaking up the components that make up the ID.  Will semver support maven snapshots?  I proposed the MavenRepositorySecret primarily to reduce the amount of boilerplate and repetition that a developer would have to go through for one artifact.  Now multiply that by a namespace that might have 10s to 100 artifacts and that might be a poor developer experience.  Given I would assume that most artifacts in a given namespace might all come from the same maven repository.

Reply:
I figure that the `classifier` field will be used to indicate that you're looking for a SNAPSHOT or not.

Reply:
`classifier` can be used in conjunction with SNAPSHOT as a part of the version. 

![image](https://user-images.githubusercontent.com/910329/163841617-0d664579-34a4-4355-a864-b577cd8b67e6.png)

In addition, in order to detect an update to a snapshot typically means that updates are made to the maven-metadata.xml associated with the artifact.  The SNAPSHOT portion of the artifact name will be replaced with a timestamp.  Here is an example of snapshots for the `spring-boot` artifact: 

https://repo.spring.io/ui/native/snapshot/org/springframework/boot/spring-boot/3.0.0-SNAPSHOT/

Notice each build of the snapshot has a timestamp replacing the SNAPSHOT placeholder in the version. The version in the k8s resource would be 3.0.0-SNAPSHOT but upon resolution it would return `spring-boot-3.0.0-20220418.130925-428.jar`.  The maven-metadata.xml can be useful hear to know what latest artifact is: 

https://repo.spring.io/artifactory/snapshot/org/springframework/boot/spring-boot/3.0.0-SNAPSHOT/maven-metadata.xml

Also, take note that the `classifier` will be called out in the metadata for `sources` and `javadoc` which are artifacts purposed at the same time but for different `classified` use cases.

Reply:
I assume that if a user is watching for new SNAPSHOT builds of some artifact then they will want to see the supply chain triggered for each individual build, right?  And each build version will have a different timestamp in it?

I will need to check if the Mastermind semver library supports strings like "SNAPSHOT" in the semver strings.  If not, then I'll need to add a boolean "is_semver" field or something similar.

Reply:
> I assume that if a user is watching for new SNAPSHOT builds of some artifact then they will want to see the supply chain triggered for each individual build, right? And each build version will have a different timestamp in it?

Yes, this is exactly what the desired outcome would be.   I imagine that the resource would poll every minute or every few minutes the [maven-metadata.xml](https://repo.spring.io/artifactory/snapshot/org/springframework/boot/spring-boot/maven-metadata.xml) at the root of the repo for a given groupId and artifactId.  This would tell me which versions are available.  This could be helpful in the case of the semver example you provided above.  If its not in the list then a folder will not exist that corresponds with that version. If the version does exist then a [maven-metadata.xml](https://repo.spring.io/artifactory/snapshot/org/springframework/boot/spring-boot/3.0.0-SNAPSHOT/maven-metadata.xml) must exist for that given version, which will contain the information such as the following for the lastUpdated version: 

```
<?xml version="1.0" encoding="UTF-8"?>
<metadata modelVersion="1.1.0">
  <groupId>org.springframework.boot</groupId>
  <artifactId>spring-boot</artifactId>
  <version>3.0.0-SNAPSHOT</version>
  <versioning>
    <snapshot>
      <timestamp>20220419.103253</timestamp>
      <buildNumber>431</buildNumber>
    </snapshot>
    <lastUpdated>20220419103350</lastUpdated>
    <snapshotVersions>
...
    </snapshotVersions>
</metadata>
```

This would apply to snapshots.  When a version does not contain -SNAPSHOT suffix then maven-metadata.xml will not exist in the version folder similar to [this](https://repo.spring.io/ui/native/libs-release/org/springframework/boot/spring-boot/2.6.6/).  However, a [maven-metadata.xml](https://repo.spring.io/artifactory/libs-release/org/springframework/boot/spring-boot/maven-metadata.xml) will exist at the groupId + artifactId level of the maven repository containing what the latest released version is based upon semver and previous versions that have been released in the past that might be candidates for a semver filter.

Reply:
Okay, so it sounds like if an end user wants to deploy SNAPSHOT builds on their cluster then the supply chain would have to go through a different flow:  it would need to look at different metadata files in different folders.  It would have to look for the builds with the newest timestamp and supply that version to the rest of the supply chain.

I'll add an `isSnapshot` boolean field to indicate that the controller will need to follow a snapshot flow.

Reply:
Oh, I understand more now.  We won't need an `isSnapshot` field to distinguish the two flows.  The presence of a -SNAPSHOT versions in in the base `metadata.xml` and individual `metadata.xml` files per version indicate the presence of SNAPSHOT builds.  The controller will be able to figure out which flow to use without an extra field.  Thank you for your explanations.  They were very helpful.

Reply:
I agree, @rszumlakowski, yes I believe the presence of -SNAPSHOT in the semver field should designate to look at the snapshotVersions section of the maven-metadata.xml

Reply:
Also, wanted to state that `packaging` should be optional and the default value would be jar.  I believe the majority of users are going to use jar packaging format. 

If semver is optional what is going to be the default behavior? 
Should the semver field be called version to be more consistent with the maven naming conventions?

Reply:
We are aiming for the name `semver` since that's what the FluxCD `GitRepository` resource uses.  We are using the `GitRepository` resource to model this behaviour.

If the semver field is left out then the controller would always return the highest version of the artifact.

I agree that the `packaging` should be optional and the default should be `jar`.

Reply:
> We are aiming for the name semver since that's what the FluxCD GitRepository resource uses. We are using the GitRepository resource to model this behaviour.

I thought a lot about this and consistency is important.  However, the audience of the MavenResource is likely to get confused by the field semver.  Other fields that exist on the resource map to their counter parts in the pom.xml (e.g. groupId, artifactId, packaging etc.)  However, semver does not match.  I propose for usability sake the semver field be named `version` to be consistent with the maven pom instead of consistent with the GitRepository resource.  Furthermore, semver for the GitRepository resource makes sense since git uses that terminology.

Reply:
The `semver` field is an option on a Flux GitRepository and not the only way to specify a commit. You can also specify a commit directly by its `commit` ref, `tag` or the HEAD of a `branch`. Curiously, Cartographer does not expose `semver` as an option for git source and nobody has asked for it yet.

The `version` field in maven has its own [range syntax](https://maven.apache.org/pom.html#Dependency_Version_Requirement_Specification).

Reply:
@scothis that is a great point on the range syntax that maven has is slightly different than the syntax that was shared above.  I would assume that keeping the syntax similar to what Maven users are familiar with.

Reply:
At Cartographer office hours today we decided to go use the Maven field name `version` and the Maven range syntax rather than the `semver` field as previously discussed.